set_test.go

```diff
@@ -147,3 +235,105 @@ func TestSetIter(t *testing.T) {
 		t.Fatalf("unexpected sum: %d, want %d", sum, sumTo99)
 	}
 }
+
+func testSetMemoryUsage[K IntKey]() func(t *testing.T) {
+	return func(t *testing.T) {
+		s := NewSet[K](16)
+		m := New[K, struct{}](16)
+		szs := int(unsafe.Sizeof(s.data[0]))
+		msz := int(unsafe.Sizeof(m.data[0]))
+		if szs*2 < msz {
+			t.Fatalf("The per-element size of Set[%T] (%d) should be half that of Map[%T] (%d)",
+				s.data[0], szs, m.data[0], msz)
+		}
+		t.Logf("Sizeof(Set[%T]) = %d; Sizeof(Map[%T]) = %d",
+			s.data[0], szs, m.data[0], msz)
+	}
+}
+
+func TestSetMemoryUsage(t *testing.T) {
+	t.Run("int", testSetMemoryUsage[int]())
+	t.Run("int8", testSetMemoryUsage[int8]())
+	t.Run("int16", testSetMemoryUsage[int16]())
+	t.Run("int32", testSetMemoryUsage[int32]())
+	t.Run("int64", testSetMemoryUsage[int64]())
+
+	t.Run("uint", testSetMemoryUsage[uint]())
+	t.Run("uint8", testSetMemoryUsage[uint8]())
+	t.Run("uint16", testSetMemoryUsage[uint16]())
+	t.Run("uint32", testSetMemoryUsage[uint32]())
+	t.Run("uint64", testSetMemoryUsage[uint64]())
+}
+
+func BenchmarkSetAdd10K(b *testing.B) {
+	rr := rand.New(rand.NewSource(12345))
+	seen := make(map[int]struct{}, 10_000)
+	for len(seen) < 10_000 {
+		seen[int(rr.Int31())] = struct{}{}
+	}
+	keys := slices.AppendSeq(make([]int, 0, 10_000), maps.Keys(seen))
+
+	s := NewSet[int](10_000)
+	b.ResetTimer()
+
+	j := 0
+	for i := 0; i < b.N; i++ {
+		s.Add(keys[j])
+		j++
+		if j == 10_000 {
+			j = 0
+			s.Clear()
+		}
+	}
+}
+
+func BenchmarkSetDel10K(b *testing.B) {
+	rr := rand.New(rand.NewSource(12345))
+
+	s := NewSet[int](10_000)
+	keys := make([]int, 0, 10_000)
+	for s.Len() < 10_000 {
+		v := rr.Int()
+		if s.Add(v) {
+			keys = append(keys, v)
+		}
+	}
+
+	// Duplicate S to make resetting it back
+	// to a populated state faster.
+	dup := *s
+	dup.data = slices.Clone(dup.data)
+
+	b.ResetTimer()
+
+	j := 0
+	for i := 0; i < b.N; i++ {
+		s.Del(keys[j])
+		j++
+		if j == 10_000 {
+			j = 0
+			b.StopTimer()
+			copy(s.data, dup.data)
+			s.size = dup.size
+			s.hasZeroKey = dup.hasZeroKey
+			b.StartTimer()
+		}
+	}
```

Comment:
This can be cleaned up a bit, and you are implicitly using the benchmark index `i` for the key to delete.

```suggestion
	b.ResetTimer()

	for range b.N {
		// Duplicate s to make resetting it back to a populated state faster.
		b.StopTimer()
		dup := *s
		dup.data = slices.Clone(dup.data)
		b.StartTimer()

		for _, k := range keys {
			dup.Del(k)
		}
	}
```